### PR TITLE
proposal for custom networks

### PR DESCRIPTION
**Pull request by DJCordhose**
hidden layers as a simple list

### CODE REVIEW

**Comment by DJCordhose** [comment]
Current API taken from tests

        custom_hidden_layers = [
            tf.keras.layers.Dense(500),
            tf.keras.layers.BatchNormalization(),
            tf.keras.layers.ReLU(),
            tf.keras.layers.Dropout(rate=0.4),
            tf.keras.layers.Dense(250),
            tf.keras.layers.BatchNormalization(),
            tf.keras.layers.ReLU(),
            tf.keras.layers.Dropout(rate=0.4)
        ]
        agent = ReinforceAgent(self.gym_env_name, training=TrainingFast(), custom_hidden_layers=custom_hidden_layers)
        agent.train()

Issues to discuss:
- This works, but uses quite a bit of copied code from tf agents
- Shall we rather pass a complete Keras model which allows for all kinds of bells and whistles while this approach only allows for simple sequential models
- How would we do this for other kinds of networks? Like q network or value network?

**Comment by christianhidber** [comment]
starting point:
o it would be nice to be able to pass an "arbitrary" keras model to any of the easyagent algorithms.
o passing "just a nn" is not sufficient, since rg dqn needs only 1 nn while actor/critc need at least 2
o different implementations (eg. tfagents) need / expect a particular form (eg qnetwork)
o currently only fc_layers is supported, which is easy to implement in all cases.

=> passing an algorithm specific, but implementation independent keras model would be great.
=> some more prototyping may be useful to establish feasability across different implementations

**Comment by DJCordhose** [comment]
> starting point:
> o it would be nice to be able to pass an "arbitrary" keras model to any of the easyagent algorithms.
> o passing "just a nn" is not sufficient, since rg dqn needs only 1 nn while actor/critc need at least 2
> o different implementations (eg. tfagents) need / expect a particular form (dqn:qnetwork vs ppo:actor/critic vs reinforce)
> 

Passing a list of sequential layers should work for all types of algorithms even with two nns just like the (500, 500, 500) since each network would be initialize them as a part of their model. 

Using a complete model instead, we would either need to get input and output right (in the existing case we just pass hidden layers) or somehow tweak the two models. If we need more than two networks we might just pass in two models? 

Sounds like an advanced use case and could coexist with this one?

> => passing an algorithm specific, but implementation independent keras model would be great.

probably not need (see above)

> => some more prototyping may be useful to establish feasability across different implementations

totally yes, it should just work with dqn and ppo, maybe we need to copy a bit of code for each of the networks, though.

> => I like it, shall we merge it into master and create an issue as a reminder to implement this for ppo/dqn as well ?

Yes.

**Comment by DJCordhose** [comment]
If we merge we would need two new issues
1. let this work with the other implementations
2. investigate arbitrary Keras models. Just hidden vs complete models?
